### PR TITLE
Issue 1611

### PR DESCRIPTION
**Pull request by davidnoahball**
## What this PR does
This addressed the feature request from issue 1611. It adds a button to the Timeline page to edit week titles, and a button to reset to default titles. I also moved dates to be below the titles to accommodate the change.

## Screenshots
Before:
Screenshot from 2019-03-18 19-50-23

After:
Screenshot from 2019-03-18 19-49-49

### CODE REVIEW

File: app/assets/javascripts/components/timeline/timeline.jsx [hunk 1]
```diff
@@ -60,6 +64,10 @@ const Timeline = createReactClass({
     return week.blocks;
   },
 
+  usingCustomTitles() {
+    return this.props.weeks.find(week => week.title) !== undefined;
```
Comment (bwreid) [inline app/assets/javascripts/components/timeline/timeline.jsx]: So, this line is trying to determine whether or not _anything_ has a title, correct? If that's the case, the following might provide a more clear intention:
```js
return this.props.weeks.some(week => week.title)
```
Alternatively, you could just return a boolean value with your `.find()` result instead of having to compare to `undefined`:
```js
return !!this.props.weeks.find(week => week.title)
```

File: app/assets/javascripts/components/timeline/timeline.jsx [hunk 2]
```diff
@@ -280,18 +296,37 @@ const Timeline = createReactClass({
...
-        <button className="button dark button--block" onClick={this.props.saveGlobalChanges}>
-          {I18n.t('timeline.save_all_changes')}
-        </button>
-        <button className="button button--clear button--block" onClick={this.props.cancelGlobalChanges}>
-          {I18n.t('timeline.discard_all_changes')}
-        </button>
+        {saveChangesButton}
+        {cancelChangesButton}
+      </div>
+    ) : undefined;
```
Comment (bwreid) [inline app/assets/javascripts/components/timeline/timeline.jsx]: This is minor, but for an intentional "nothing" value, I would typically use `null`.

File: app/assets/javascripts/components/timeline/timeline.jsx [hunk 3]
```diff
@@ -280,18 +296,37 @@ const Timeline = createReactClass({
...
+    const titlesActionButtons = this.props.editableTitles ? (
+      <div>
+        {saveChangesButton}
+        {cancelChangesButton}
+        {resetTitlesButton}
       </div>
     ) : undefined;
```
Comment (bwreid) [inline app/assets/javascripts/components/timeline/timeline.jsx]: Same goes for here.

File: app/assets/javascripts/reducers/timeline.js [hunk 4]
```diff
@@ -39,6 +41,13 @@ const blocksInWeek = (blocks, weekId) => {
   return count;
 };
 
+const validateTitle = (title) => {
+  if ((typeof title === 'string' || title instanceof String) && title.length <= 20) {
+    return true;
+  }
+  return false;
+};
```
Comment (bwreid) [inline app/assets/javascripts/reducers/timeline.js]: You can just shorten this to:
```js
return (typeof title === 'string' || title instanceof String) && title.length <= 20)
```

What's the case where this would not be a string?

File: app/assets/stylesheets/modules/_timeline.styl [hunk 5]
```diff
@@ -191,7 +191,7 @@ button.week__add-block
 
 .week__week-dates
   position relative
-  top 16px
+  top -12px
```
Comment (bwreid) [inline app/assets/stylesheets/modules/_timeline.styl]: If this is going to appear right below the title, I think it would be better to rely on more standard techniques like `margin`. We probably don't need to reposition this anymore.

**Comment by davidnoahball** [comment]
@bwreid Thanks for the review. I've addressed your comments and fixed the bug. I think it should be good to go now.

**Comment by bwreid** [comment]
@davidnoahball Thanks so much! I think this looks great.

I would like to wait until @ragesoss gets back so he can also weigh in on the UI changes. As soon as he does so, we'll merge it. Great work!

File: app/assets/javascripts/components/timeline/week.jsx [hunk 6]
```diff
@@ -68,21 +71,35 @@ const Week = createReactClass({
...
+    } else {
+      weekTitleContent = I18n.t('timeline.week_number', { number: this.weekNumber() });
+    }
+    const weekId = this.props.week.id;
+    const weekTitle = this.props.editableTitles ? (
+      <input
+        className="week-index week-title-input"
+        defaultValue={weekTitleContent}
+        maxLength={20}
+        onChange={text => this.props.updateTitle(weekId, event.target.value, text)}
```
Comment (ragesoss) [inline app/assets/javascripts/components/timeline/week.jsx]: This is giving me an error. Should be `event => this.props.updateTitle(weekId, event.target.value)`?

Comment (ragesoss) [reply app/assets/javascripts/components/timeline/week.jsx]: This error just happens in Firefox, it looks like.

**Comment by ragesoss** [comment]
The Milestones thing doesn't need to be done in this PR; there's already a separate issue for it. This should be good to merge once the `onChange` is fixed and the button label is updated.

**Comment by davidnoahball** [comment]
@ragesoss `onChange` fixed, button label updated

**Comment by davidnoahball** [comment]
Forgot about updating the button in the test; fixed now.
